pkg/csi/service/wcpguest/controller.go

```diff
@@ -420,6 +453,128 @@ func (c *controller) ControllerPublishVolume(ctx context.Context, req *csi.Contr
+				Name:      cnsFileAccessConfigInstanceName,
+				Namespace: c.supervisorNamespace},
+			Spec: cnsfileaccessconfigv1alpha1.CnsFileAccessConfigSpec{
+				VMName:  req.NodeId,
+				PvcName: req.VolumeId,
+			},
+		}
+		log.Debugf("Creating CnsFileAccessConfig instance: %+v", cnsFileAccessConfigInstance)
+		if err := c.cnsOperatorClient.Create(ctx, cnsFileAccessConfigInstance); err != nil {
+			msg := fmt.Sprintf("failed to create cnsFileAccessConfig: %v. Error: %v", cnsFileAccessConfigInstance.Name, err)
+			log.Error(msg)
+			return nil, status.Errorf(codes.Internal, msg)
+		}
+	}
+	log.Debugf("Found CnsFileAccessConfig: %q", cnsFileAccessConfigInstance.Name)
+	if cnsFileAccessConfigInstance.DeletionTimestamp != nil {
+		// When deletionTimestamp is set, CnsOperator is in the process of removing access for this IP.
+		// When that operation is successful, the instance will be deleted. In a subsequent retry, a new instance will be created.
+		msg := fmt.Sprintf("cnsFileAccessConfigInstance %q is getting deleted. A new instance will be created in the subsequent ControllerPublishVolume request", cnsFileAccessConfigInstance.Name)
+		log.Error(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+	// Verify if the CnsFileAccessConfig instance has status with done set to true and error is empty
+	if cnsFileAccessConfigInstance.Status.Done && cnsFileAccessConfigInstance.Status.Error == "" {
+		publishInfo := make(map[string]string)
+		for key, value := range cnsFileAccessConfigInstance.Status.AccessPoints {
+			if key == common.Nfsv4AccessPointKey {
+				publishInfo[common.Nfsv4AccessPoint] = value
+				break
+			}
+		}
+		resp := &csi.ControllerPublishVolumeResponse{
+			PublishContext: publishInfo,
+		}
+		log.Infof("ControllerPublishVolume: Volume %q attached successfully on the node: %q", req.VolumeId, req.NodeId)
+		return resp, nil
+	}
+	cnsFileAccessConfigWatcher, err := k8s.NewCnsFileAccessConfigWatcher(ctx, c.restClientConfig, c.supervisorNamespace)
+	if err != nil {
+		msg := fmt.Sprintf("failed to create cnsFileAccessConfigWatcher. Error: %+v", err)
+		log.Error(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+	// Attacher timeout, default is set to 4 minutes
+	timeoutSeconds := int64(getAttacherTimeoutInMin(ctx) * 60)
+	// Adding watch on the CnsFileAccessConfig instance to register for updates
+	watchCnsFileAccessConfig, err := cnsFileAccessConfigWatcher.Watch(metav1.ListOptions{
+		FieldSelector:   fields.SelectorFromSet(fields.Set{"metadata.name": cnsFileAccessConfigInstance.Name}).String(),
+		ResourceVersion: cnsFileAccessConfigInstance.ResourceVersion,
+		TimeoutSeconds:  &timeoutSeconds,
+	})
+	if err != nil {
+		msg := fmt.Sprintf("failed to watch cnsfileaccessconfig %q with Error: %v", cnsFileAccessConfigInstance.Name, err)
+		log.Error(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+	defer watchCnsFileAccessConfig.Stop()
+	timeout := time.Now().Add(time.Duration(timeoutSeconds) * time.Second)
```

Comment:
Since the timeout for watch is already set on line 532, is this needed?

Reply:
This is unnecessary. Removed